### PR TITLE
Basic tei transcription conversion & search

### PR DESCRIPTION
**Pull request by rlskoeser**
Adds a new flask command to do basic parsing of TEI transcriptions that creates a json file with extracted information, updates indexing to include transcription data when available, adds transcription to search fields, and adds highlighting of matches in transcription text.

To generate transcription data file:
1. Download a local copy of the TEI/XML data
2. Configure the path in flask local settings as **XML_TRANSCRIPTIONS_DIR**
3. Run the command: `flask transcriptions`

Once you have the transcription data file, you can run `flask index` to index metadata + transcription text.

Requires a solr config change to search on transcription text.

### CODE REVIEW

File: scripts/index.py [hunk 1]
```diff
@@ -58,7 +76,9 @@ def index():
             'link_s': extlink or None,
             'iiif_link_s': iiif_link,
             'editors_txt': row['Editor(s)'] or None,
-            'translators_txt': row['Translator (optional)'] or None
+            'translators_txt': row['Translator (optional)'] or None,
```
Comment (kmcelwee) [inline scripts/index.py:79]: oh yeah this is certainly more straightforward than what I was doing 🤔

**Review by kmcelwee — APPROVED** [review]

File: scripts/tei_transcriptions.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                'label': '\n'.join(label),
+                'lines': lines,
+                'languages': list(languages)
+            })
+
+        docdata = {
+            'blocks': blocks,
+            'lines': [str(l) for l in tei.text.lines]
+        }
+        data[tei.pgpid] = docdata
```
[the remark this answers is not in the document]
Comment (rlskoeser) [reply scripts/tei_transcriptions.py:82]: If we weren't writing experimental code, we would write unit tests to confirm that the logic is sound and handles known cases, including errors. And if we get to a point in the experimental code where unit tests would be helpful, we will write them!

File: scripts/index.py [hunk 3]
```diff
@@ -45,9 +49,23 @@ def index():
         if 'cudl.lib.cam.ac.uk' in extlink:
             iiif_link = extlink.replace('/view/', '/iiif/')
 
+        pgpid = row['PGPID']
+        text = text_blob = None
+        if pgpid in transcriptions:
+            # TBD: should labels also be searchable ?
+            # throw everything into a text blob
+            text_blob = ' '.join(transcriptions[pgpid]['lines'])
```
[the remark this answers is not in the document]
Comment (rlskoeser) [reply scripts/index.py:57]: I don't think the whitespace matters — Solr will ignore it when the text is tokenized. Whitespace _does_ matter in the actual text we're displaying; currently I'm not displaying this version, but inclined not to strip it until we know whether it matters to us.

File: .gitignore [hunk 4]
```diff
@@ -1,5 +1,6 @@
 # local settings
 scripts/local_settings.py
+data/transcriptions.json
```
Comment (thatbudakguy) [inline .gitignore:3]: we could probably ignore an entire `data/` directory; the csv files and other things I've downloaded for the NER experiment would go there to be ignored

File: scripts/index.py [hunk 5]
```diff
@@ -45,9 +49,23 @@ def index():
...
+            # but also index as lines so highlighting can return single lines
+            # instead of the whole text
+            text = transcriptions[pgpid]['lines']
+            # index language from spreadsheet?
+            # TODO: languages will need to be auto-detected; may be useful
+            # to check against language from the spreadsheet
```
Comment (thatbudakguy) [inline scripts/index.py:63]: Should we create an experiment issue to track trying this out?

File: scripts/tei_transcriptions.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    text = xmlmap.NodeField('tei:text/tei:body/tei:div|tei:text/tei:body[not(tei:div)]', MainText)
+    lines = xmlmap.NodeListField('tei:text/tei:body/tei:div/tei:l',
+                                 GenizaTeiLine)
+    labels = xmlmap.NodeListField('tei:text/tei:body/tei:div/tei:label',
+                                  GenizaTeiLine)   # not really a line...
+
+
+@click.command()
+@with_appcontext
+def transcriptions():
```
Comment (thatbudakguy) [inline scripts/tei_transcriptions.py:35]: not super important, but i might give this a slightly more descriptive name, like `transcriptions_to_json()` or something

File: scripts/server.py [hunk 7]
```diff
@@ -36,13 +37,27 @@ def search():
...
         queryset = queryset.search(search_query) \
             .raw_query_parameters(search_terms=search_terms) \
+            .highlight('transcription_lines_txt', snippets=3, method='unified') \
             .order_by('-score').only('*', 'score')
 
     results = queryset.get_results(rows=50)
 
+    # copied from pemm
+    if results and search_terms:
+        # patch in the highlighted incipits into the main result
```
Comment (thatbudakguy) [inline scripts/server.py:52]: not incipits anymore :)

File: scripts/tei_transcriptions.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+                    })
+                    label = []
+                    lines = []
+
+                # store the label; sometimes there are two in a row
+                label.append(str(line))
+
+            elif line.name == 'l':
+                if line.lang:
+                    # TODO: can we assume a default/unspecified language where not set?
```
Comment (thatbudakguy) [inline scripts/tei_transcriptions.py:67]: Is this a question for the team?

**Review by thatbudakguy — APPROVED** [review]
looks good, sorry it took me a bit to review!

File: scripts/index.py (hunk 5 above)
Comment (rlskoeser) [reply scripts/index.py:63]: @thatbudakguy yes, good point. I've been meaning to create one so I can document the spacy library you recommended for this.

File: scripts/tei_transcriptions.py (hunk 6 above)
Comment (rlskoeser) [reply scripts/tei_transcriptions.py:35]: The function name corresponds to the flask command name, so I'm inclined to leave it shorter (already kind of long). We can always rename when/if we have more transcription scripts.

File: scripts/tei_transcriptions.py (hunk 8 above)
Comment (rlskoeser) [reply scripts/tei_transcriptions.py:67]: Ah. I did ask them! Hence the note about needing to detect languages. I will revise this comment!

**Pull request merged**